### DISCUSSION TITLE
UEFI-based laptops not displaying the iPXE boot menu

### DISCUSSION THREAD

**Discussion by jamaya77** [ipxe/ipxe#290] [Q&A]
I'm a newbie and testing iPXE to replace the old PXE solution. For my test environment, I'm using CentOS 8 with iPXE, Dnsmasq and Apache installed from the repository. Dnsmasq is configured as a dhcpProxy because of an existing server.

iPXE appears to be working as both UEFI and BIOS-based laptops are able to pxe boot with iPXE. However, I can only get bios-based laptops to load to the boot menu.

Please advise. Thank you.

This my dnsmasq configuration:
#disable dns server
port=0

#enable tftp server
enable-tftp tftp-root=/tftpboot

#dhcp proxy mode. do not act as the authoratitive dhcp server
dhcp-range=192.168.140.0,proxy,255.255.252.0

#send disable multicast and broadcast discovery, and to download the boot file immediately
dhcp-option=vendor:PXEClient,6,2b

#set tag ENH if request comes from iPXE (iPXE user class)
dhcp-userclass=set:ENH,iPXE

#if request comes from firmware, load iPXE via TFTP
pxe-service=tag:!ENH,X86-64_EFI, UEFI: Chainload Boot - iPXE, ipxe.efi

pxe-service=tag:!ENH,X86PC, PC-DOS: Chainload Boot - iPXE, undionly.kpxe

#if request comes from iPXE, direct it to boot from boot.ipxe
pxe-service=tag:ENH,X86-64_EFI, UEFI: ipxe.efi - iPXE, http://192.168.140.21/menu/boot.ipxe

pxe-service=tag:ENH,X86PC, PC-DOS: undionly.kpxe - iPXE, http://192.168.140.21/menu/boot.ipxe

#2nd boot file
dhcp-boot=tag:ENH,http://192.168.140.21/menu/boot.ipxe

#enable logging
log-queries

log-dhcp

<img width="1137" alt="ipxeuefi" src="https://user-images.githubusercontent.com/80473349/110826608-40795a00-8263-11eb-9b5b-52d5fbc6c248.png">

**Comment by mcb30** [comment]
You say that both UEFI and BIOS are able to boot, but you also say that only BIOS is able to get to the boot menu.  The immediately obvious question is: what happens when you try to boot the UEFI-based laptop?

When submitting any problem report, it's always a good idea to say what it is that actually happens! :slightly_smiling_face:

**Reply by jamaya77** [reply]
Hi mcb30,
Thank you for your response. I've added a snapshot.

**Reply by NiKiZe** [reply]
in that screenshot we can see that iPXE runs and downloads ipxe.efi, so the issue here is that iPXE gets the ipxe.efi filename instead of the expected http://192.168.140.21/menu/boot.ipxe

I find the dnsmasq logic hard to follow, looking at my own config, I ended up never getting it to work, and instead used EMBED with below script:

```ipxe
#!ipxe
set esc:hex 1b && set esc ${esc:string}[
set cls ${esc}2J
cpuid --ext 29 && set arch amd64 || set arch x86
ifstat
dhcp || shell
set nloc ${netX/busloc}
echo DNS: ${dns}
chain http://pxe.local/main.ipxe?mac=${netX/mac:hexraw}&bus=${netX/busid:hexhyp}&loc=${nloc}&id=${pci/${nloc}.0.2}${pci/${nloc}.2.2}-${pci/${nloc}.0x2c.2}${pci/${nloc}.0x2e.2}&chip=${netX/chip:uristring}&ip=${netX/ip}&next-server=${next-server}&arch=${arch}&platform=${platform:uristring}&class=${user-class:uristring}&manuf=${manufacturer:uristring}&prod=${product:uristring}&sn=${serial:uristring}&boardsn=${board-serial:uristring}&asset=${asset:uristring}&uuid=${uuid:uuid}&ver=${version:uristring}&file=${filename:uristring} || shell
```

**Reply by mcb30** [reply]
Thanks.  Your screenshot shows that iPXE has started up successfully, and that your DHCP or ProxyDHCP server has instructed it to re-download `ipxe.efi` instead of `http://192.168.140.21/menu/boot.ipxe`.

First, press Ctrl-B to enter the iPXE shell and then use `dhcp` and `show filename` to find out whether it is the DHCP server or ProxyDHCP server that is supplying the filename.

Assuming that the filename is coming as I think you expect from the ProxyDHCP server (running `dnsmasq`) then the problem must be within your `dnsmasq` configuration file.  You seem to have some redundancy between `pxe-service` and `dhcp-boot` directives: I'd probably start by cutting out the unnecessary `pxe-service` lines and using just `dhcp-boot` to provide the `http://192.168.140.21/menu/boot.ipxe` boot URL to iPXE.  Once the configuration is simplified, retest and check the various logs (and/or wireshark captures) to verify that the server is handing out the filenames that you expect at each stage.

**Reply by jamaya77** [reply]
Thank you so much for your time, guys. 

To mcb30:
The iPXE server is setup as a proxyDHCP since there's another dhcp server on the same network. The other dhcp server is configured to call for the "ipxe.efi" file which will be commented out and tested afterwards. I've also attached snapshot.
<img width="510" alt="Screen Shot 2021-03-12 at 1 02 09 AM" src="https://user-images.githubusercontent.com/80473349/110899584-5de40d80-82cf-11eb-9f63-7fb9a08d3dfd.png">

For NiKiZe:
I'd like to try your suggestion. Once I create the script, can you provide the syntax or how to embed the script to the dnsmasq.conf file?

Thanks again.

**Reply by mcb30** [reply]
A filename provided by the DHCP server will take priority over a filename provided by the ProxyDHCP server.  This is in accordance with the PXE specification (if not necessarily in accordance with the way that Intel PXE ROMs behave).

To allow the filename provided by your ProxyDHCP server to be used, you will need to reconfigure your DHCP server to not provide any boot filename.

**Comment by jamaya77** [comment]
It works!!! After commenting out filename "ipxe.efi" from the dhcp server, the loop stopped and the loaded the boot menu for both uefi and bios laptops. Thank you! I truly appreciate the fast response, guys.

The issue is resolved.
